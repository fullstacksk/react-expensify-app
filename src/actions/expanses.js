
import { v4 as uuid } from 'uuid';
import database from '../firebase/firebase';
// ADD_EXPANSE
export const addExpanse = (expanse) => (
    {
        type: 'ADD_EXPANSE',
        expanse
    }
);
export const startAddExpanse = (expanseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expanseData;
        const expanse = { description, note, amount, createdAt };
        return database.ref('expanses')
            .push(expanse)
            .then((ref) => {
                dispatch(addExpanse({
                    id: ref.key,
                    ...expanse
                }))
            })
            .catch((e) => {
                console.log("Add Expanse Failed", e)
            })
    }
}

// REMOVE_EXPANSE
export const removeExpanse = ({ id = '' } = {}) => ({
    type: 'REMOVE_EXPANSE',
    id

});
// EDIT_EXPANSE
export const editExpanse = (id, updates) => ({
    type: 'EDIT_EXPANSE',
    id,
    updates
});