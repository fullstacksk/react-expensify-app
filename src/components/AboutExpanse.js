import React from 'react';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Alert from 'react-bootstrap/Alert';

const AboutExpanse = () => (
    <Container>
        <Container>
            <Alert variant="success">
                <Row className="px-2">
                    <h5>About Expanse Manager</h5>
                    <p>Visual Studio Code and WebStorm support debugging out of the box with Create React App. This enables you as a developer to write and debug your React code without leaving the editor, and most importantly it enables you to have a continuous development workflow, where context switching is minimal, as you donâ€™t have to switch between tools.</p>
                </Row>
            </Alert>
        </Container>
    </Container>
);

export default AboutExpanse;