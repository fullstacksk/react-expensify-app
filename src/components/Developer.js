import React from 'react';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Alert from 'react-bootstrap/Alert';

const Developer = () => (
    <Container>
        <Container>
            <Alert variant="success" >
                <Row className="px-2">
                    <Col>
                        <h5>Fullstack Developer</h5>
                        <p>Hey, I am <code>SK Sahil</code> and this was the my first <kbd>React</kbd> web app.</p>
                        <h6>About React Js</h6>
                        <p>React Js is an wonderful js library <code>not a framework</code> which is  being used for dynamic front end development. It is developed by <code>Facebook</code> itself. It uses nodejs internally as programming language. NodeJs works on chromes <kbd>V8 Engine</kbd></p>
                    </Col>
                </Row>
            </Alert>
            <Alert variant="info" >
                <Row className="px-2">
                    <h5>Connect with me</h5>
                </Row>
                <Row>
                    <Col md={4}>
                        <a className="btn btn-primary" href="https://www.facebook.com/profile.php?id=100008404923891">Facebook</a>
                    </Col>
                    <Col md={4}>
                        <a className="btn btn-warning" href="https://www.instagram.com/sk_sahil_official/" >Instagram</a>
                    </Col>
                    <Col md={4}>
                        <a className="btn btn-secondary" href="https://github.com/fullstacksk/react-expensify-app">Github</a>
                    </Col>
                </Row>
            </Alert>
            <Alert variant="info">
                <Row>
                    <Col>
                        <h6>Suggest Me At</h6>
                        <kbd>fullstacksk@gmail.com</kbd></Col>
                </Row>
            </Alert>
        </Container>
    </Container >
);

export default Developer;