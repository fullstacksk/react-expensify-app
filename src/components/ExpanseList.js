import React from 'react';
import moment from 'moment';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import slectExpanses from '../slectors/expanses';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Badge from 'react-bootstrap/Badge';
import Alert from 'react-bootstrap/Alert';

export const ExpanseList = (props) => (
    <Container>
        <Container>
            <Alert variant="warning" className="mb-1 py-2">
                <Row className="justify-content-between px-2">
                    <div><h5>Expanse<Badge variant="success" className="ml-2">{props.expanses.length}</Badge></h5></div>
                    <div><h5>Amount<Badge variant="warning" className="ml-2">
                        {
                            props.expanses.length ?
                                props.expanses.map(
                                    ({ amount = '0' }) => parseFloat(amount)
                                ).reduce((a = 0, v = 0) => a + v)
                                : 0
                        }
                    </Badge></h5></div>
                </Row>
            </Alert>
        </Container>
        {
            props.expanses.length ?

                <Container>
                    {
                        props.expanses.map((expanse, idx) => (
                            <NavLink
                                to={`/edit/${expanse.id}`}
                                key={idx}
                            >

                                <Alert variant="info" className="mb-2 shadow-sm ">
                                    <Row className="justify-content-between px-2">
                                        <div>
                                            <h5 className="pb-0">{expanse.description}</h5>
                                            <span className="small mt-0">{moment(expanse.createdAt).format('Do MMM, YYYY')}</span>
                                        </div>
                                        <div><h5>{expanse.amount}</h5></div>
                                    </Row>
                                </Alert>
                            </NavLink>
                        ))
                    }
                </Container >
                :

                <Container>
                    <Alert variant="dark" className="mb-1 py-2">
                        <Row className="justify-content-center px-2">
                            <div><h6><Badge variant="success" className="ml-2">No</Badge> Expanses</h6></div>
                        </Row>
                    </Alert>
                </Container>
        }
    </Container >
);
const mapStateToProps = (state) => ({
    expanses: slectExpanses(state.expanses, state.filters),
    filters: state.filters
});
export default connect(mapStateToProps)(ExpanseList);