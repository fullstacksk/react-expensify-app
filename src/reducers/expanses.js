
//Expanses Reducer
const expansesReducerDefaultState = [];

export default ((state = expansesReducerDefaultState, action) => {
    switch (action.type) {
        case "SET_EXPANSES":
            return action.expanses
        case 'ADD_EXPANSE':
            return [
                ...state,
                action.expanse
            ];
        case 'REMOVE_EXPANSE':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPANSE':
            return state.map((expanse) => {
                if (expanse.id === action.id)
                    return {
                        ...expanse,
                        ...action.updates
                    }
                return expanse;
            });
        default:
            return state;
    }
});
