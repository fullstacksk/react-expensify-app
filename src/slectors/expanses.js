import moment from 'moment';
//get visible expanses
export default (expanses, { text, sortBy, startDate, endDate }) => {
    return expanses.filter((expanse) => {
        const craetedAtMoment = moment(expanse.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(craetedAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(craetedAtMoment, 'day') : true;
        const textMatch = expanse.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            // console.log(a, b);
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    })
}
