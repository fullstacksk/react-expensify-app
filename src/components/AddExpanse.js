import React from 'react';
import ExpanseForm from './ExpanseForm';
import { addExpanse } from '../actions/expanses';
import { connect } from 'react-redux';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Alert from 'react-bootstrap/Alert';

const AddExpanse = (props) => (
    <Container>
        <Container>
            <Alert variant="info">
                <Row className="text-success px-2">
                    <h5>You can add your expanse!</h5>
                </Row>
            </Alert>
        </Container>
        <ExpanseForm
            onSubmit={(expanse) => {
                props.dispatch(addExpanse(expanse));
                props.history.push("/");
            }}
        />
    </Container>
);

export default connect()(AddExpanse);