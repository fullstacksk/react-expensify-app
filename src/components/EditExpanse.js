import React from 'react';
import { connect } from 'react-redux';
import { startEditExpanse, startRemoveExpanse } from '../actions/expanses';
import ExpanseForm from './ExpanseForm';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Alert from 'react-bootstrap/Alert';

export class EditExpanse extends React.Component {

    onSubmit = (expanse) => {
        this.props.startEditExpanse(this.props.expanse.id, expanse);
        this.props.history.push('/dashboard');
    };
    startRemoveExpanse = () => {
        this.props.startRemoveExpanse(this.props.expanse.id);
        this.props.history.push('/dashboard');
    }
    render() {
        return (
            <Container>
                <Container className="pt-3 pb-1">
                    <Alert variant='info'>
                        <Row className="px-2">
                            <p className="text-success mb-0">There is a request to edit the expanse having id : <kbd className="bg-info">{this.props.expanse.id}</kbd></p>
                        </Row>
                    </Alert>
                </Container>
                <ExpanseForm

                    onSubmit={this.onSubmit}
                    expanse={this.props.expanse}
                    startRemoveExpanse={this.startRemoveExpanse}
                />
            </Container>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expanse: state.expanses.find((expanse) => expanse.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpanse: (id, expanse) => dispatch(startEditExpanse(id, expanse)),
    startRemoveExpanse: (id) => dispatch(startRemoveExpanse({ id }))
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpanse);