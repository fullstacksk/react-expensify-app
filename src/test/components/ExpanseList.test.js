import React from 'react'
import { shallow } from 'enzyme'
import { ExpanseList } from '../../components/ExpanseList'
import expanses from '../fixtures/expanses'

test("Should render ExpanseList with expanses", () => {
    const wrapper = shallow(<ExpanseList expanses={expanses} />)
    expect(wrapper).toMatchSnapshot()
})

test("Should render ExpanseList with no expanse", () => {
    const wrapper = shallow(<ExpanseList expanses={[]} />)
    expect(wrapper).toMatchSnapshot()
})