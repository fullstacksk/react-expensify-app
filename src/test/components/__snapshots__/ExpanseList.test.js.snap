// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpanseList with expanses 1`] = `
<Container
  fluid={false}
>
  <Container
    fluid={false}
  >
    <Alert
      className="mb-1 py-2"
      closeLabel="Close alert"
      show={true}
      transition={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "appear": false,
            "in": false,
            "mountOnEnter": false,
            "timeout": 300,
            "unmountOnExit": false,
          },
          "displayName": "Fade",
          "render": [Function],
        }
      }
      variant="warning"
    >
      <Row
        className="justify-content-between px-2"
        noGutters={false}
      >
        <div>
          <h5>
            Expanse
            <Badge
              className="ml-2"
              pill={false}
              variant="success"
            >
              3
            </Badge>
          </h5>
        </div>
        <div>
          <h5>
            Amount
            <Badge
              className="ml-2"
              pill={false}
              variant="warning"
            >
              6000
            </Badge>
          </h5>
        </div>
      </Row>
    </Alert>
  </Container>
  <Container
    fluid={false}
  >
    <NavLink
      key="0"
      to="/edit/123"
    >
      <Alert
        className="mb-2 shadow-sm "
        closeLabel="Close alert"
        show={true}
        transition={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "appear": false,
              "in": false,
              "mountOnEnter": false,
              "timeout": 300,
              "unmountOnExit": false,
            },
            "displayName": "Fade",
            "render": [Function],
          }
        }
        variant="info"
      >
        <Row
          className="justify-content-between px-2"
          noGutters={false}
        >
          <div>
            <h5
              className="pb-0"
            >
              Gas bill
            </h5>
            <span
              className="small mt-0"
            >
              1st Jan, 1970
            </span>
          </div>
          <div>
            <h5>
              3000
            </h5>
          </div>
        </Row>
      </Alert>
    </NavLink>
    <NavLink
      key="1"
      to="/edit/1234"
    >
      <Alert
        className="mb-2 shadow-sm "
        closeLabel="Close alert"
        show={true}
        transition={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "appear": false,
              "in": false,
              "mountOnEnter": false,
              "timeout": 300,
              "unmountOnExit": false,
            },
            "displayName": "Fade",
            "render": [Function],
          }
        }
        variant="info"
      >
        <Row
          className="justify-content-between px-2"
          noGutters={false}
        >
          <div>
            <h5
              className="pb-0"
            >
              Rent bill
            </h5>
            <span
              className="small mt-0"
            >
              28th Dec, 1969
            </span>
          </div>
          <div>
            <h5>
              1000
            </h5>
          </div>
        </Row>
      </Alert>
    </NavLink>
    <NavLink
      key="2"
      to="/edit/12345"
    >
      <Alert
        className="mb-2 shadow-sm "
        closeLabel="Close alert"
        show={true}
        transition={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "appear": false,
              "in": false,
              "mountOnEnter": false,
              "timeout": 300,
              "unmountOnExit": false,
            },
            "displayName": "Fade",
            "render": [Function],
          }
        }
        variant="info"
      >
        <Row
          className="justify-content-between px-2"
          noGutters={false}
        >
          <div>
            <h5
              className="pb-0"
            >
              Water bill
            </h5>
            <span
              className="small mt-0"
            >
              5th Jan, 1970
            </span>
          </div>
          <div>
            <h5>
              2000
            </h5>
          </div>
        </Row>
      </Alert>
    </NavLink>
  </Container>
</Container>
`;

exports[`Should render ExpanseList with no expanse 1`] = `
<Container
  fluid={false}
>
  <Container
    fluid={false}
  >
    <Alert
      className="mb-1 py-2"
      closeLabel="Close alert"
      show={true}
      transition={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "appear": false,
            "in": false,
            "mountOnEnter": false,
            "timeout": 300,
            "unmountOnExit": false,
          },
          "displayName": "Fade",
          "render": [Function],
        }
      }
      variant="warning"
    >
      <Row
        className="justify-content-between px-2"
        noGutters={false}
      >
        <div>
          <h5>
            Expanse
            <Badge
              className="ml-2"
              pill={false}
              variant="success"
            >
              0
            </Badge>
          </h5>
        </div>
        <div>
          <h5>
            Amount
            <Badge
              className="ml-2"
              pill={false}
              variant="warning"
            >
              0
            </Badge>
          </h5>
        </div>
      </Row>
    </Alert>
  </Container>
  <Container
    fluid={false}
  >
    <Alert
      className="mb-1 py-2"
      closeLabel="Close alert"
      show={true}
      transition={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "appear": false,
            "in": false,
            "mountOnEnter": false,
            "timeout": 300,
            "unmountOnExit": false,
          },
          "displayName": "Fade",
          "render": [Function],
        }
      }
      variant="dark"
    >
      <Row
        className="justify-content-center px-2"
        noGutters={false}
      >
        <div>
          <h6>
            <Badge
              className="ml-2"
              pill={false}
              variant="success"
            >
              No
            </Badge>
             Expanses
          </h6>
        </div>
      </Row>
    </Alert>
  </Container>
</Container>
`;
